chore: add Uptrace integration

vmihailenco opened the pull request:
@lanceleonard as discussed, this an early preview to gather initial feedback.

I think I already have most of the details in the doc, but there is an unfinished section and I will iterate on it tomorrow. I will probably also add another screenshot or two.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RedisLabs/redislabs-docs?pullRequest=2148) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RedisLabs/redislabs-docs?pullRequest=2148) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RedisLabs/redislabs-docs?pullRequest=2148) it.</sub>

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,185 @@
+
+![uptrace-redis-nodes](/images/rs/uptrace-redis-nodes.png)
+
+## Installing Collector and Uptrace
```
```suggestion
## Install Collector and Uptrace
```

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,185 @@
+
+Because installing OpenTelemetry Collector and Uptrace can take some time, you can use the [docker-compose](https://github.com/uptrace/uptrace/tree/master/example/redis-enterprise) example that also comes with Redis Enterprise cluster and AlertManager.
+
+Alternatively, you can try the following guides to explore all available options:
+
+<!-- The first link could point to the official Otel doc, but it focuses on installing otelcol instead of otelcol-contrib which is required here -->
```
```suggestion
Alternative installation guides include:
```

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,185 @@
+- [Getting started with OpenTelemetry Collector](https://uptrace.dev/opentelemetry/collector.html)
+- [Getting started with Uptrace](https://uptrace.dev/get/install.html)
+
+Either way, after installing Uptrace you should have Uptrace UI available at [http://localhost:14318/](http://localhost:14318/).
```
```suggestion
After you install Uptrace, you can access the Uptrace UI at [http://localhost:14318/](http://localhost:14318/).
```

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,185 @@
+
+Either way, after installing Uptrace you should have Uptrace UI available at [http://localhost:14318/](http://localhost:14318/).
+
+## Scraping Prometheus metrics
```
We also might want to consider changing the **Scrape Prometheus metrics** instructions to a numbered list of steps.
```suggestion
## Scrape Prometheus metrics
```

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,185 @@
+
+You can also check the full OpenTelemetry Collector config [here](https://github.com/uptrace/uptrace/blob/master/example/redis-enterprise/otel-collector.yaml).
+
+## Viewing metrics
```
```suggestion
## View metrics
```

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,185 @@
+
+When metrics start arriving to Uptrace, you should see a couple of dashboards in the Metrics tab. In total, Uptrace should create 3 dashboards for Redis Enterprise metrics:
+
+- "Redis: Nodes" dashboard displays a list of cluster nodes with ability to click on a node to view node metrics.
```
```suggestion
- "Redis: Nodes" dashboard displays a list of cluster nodes. You can select a node to view its metrics.
```

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,185 @@
+
+- "Redis: Databases" displays a list of Redis databases in all cluster nodes. To find a specific database, you can use filters or sort the table by columns.
+
+- "Redis: Shards" contains a list of shards that you have in all cluster nodes. Just like with other dashboards, you can filter/sort shards and click on a shard to get more details.
```
```suggestion
- "Redis: Shards" contains a list of shards that you have in all cluster nodes. You can filter or sort shards and select a shard for more details.
```

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,185 @@
+
+- "Redis: Shards" contains a list of shards that you have in all cluster nodes. Just like with other dashboards, you can filter/sort shards and click on a shard to get more details.
+
+## Monitoring metrics
```
```suggestion
## Monitor metrics
```

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
````diff
@@ -0,0 +1,185 @@
+      projects: [1]
+```
+
+## Conclusion
+
+TODO: some final words and links to the blog etc
````
I think it would be okay to remove the conclusion section unless you want to replace it with a **More info** section and add a few useful links.
```suggestion
```

rrelledge reviewed (changes requested):
Hi, @vmihailenco. Thank you for creating this new article. I just left a couple of suggestions for you to consider.

I do have a few questions about the config files. From the alternative installation guides, I see that the Uptrace config path is /etc/uptrace/uptrace.yml, but I didn't see the location of otel-collector.yaml.

1. Where is the path to otel-collector.yaml?
2. Does the docker-compose installation method add these config files to the same location as the alternative installation methods?

vmihailenco commented:
>I just left a couple of suggestions for you to consider.

Thanks, I've applied all of them and also added a new section in the end of the doc. Please take a look.

>Where is the path to otel-collector.yaml?

I've referred to the file that comes with the docker example in https://github.com/uptrace/uptrace/tree/master/example/redis-enterprise.  Inside docker container it can be found at /etc/otelcol-contrib/config.yaml

>Does the docker-compose installation method add these config files to the same location as the alternative installation methods?

The location is the same, but you can't edit files inside the docker container and that is why I have not specified the full path initially. But I guess both options have some merits so I've updated the doc to contain a full path

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,189 @@
+
+## Install Collector and Uptrace
+
+Because installing OpenTelemetry Collector and Uptrace can take some time, you can use the [docker-compose](https://github.com/uptrace/uptrace/tree/master/example/redis-enterprise) example that also comes with Redis Enterprise cluster and Alertmanager.
```
It looks like the Uptrace docs use "AlertManager" rather than "Alertmanager".
```suggestion
Because installing OpenTelemetry Collector and Uptrace can take some time, you can use the [docker-compose](https://github.com/uptrace/uptrace/tree/master/example/redis-enterprise) example that also comes with Redis Enterprise cluster and AlertManager.
```

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,189 @@
+
+## Send notifications
+
+Uptrace does not manage notifications by itself and instead provides an [integration](https://uptrace.dev/get/alerting.html) with Alertmanager.
+
+Alertmanager handles alerts sent by client applications such as Uptrace and takes care of deduplicating, grouping, and routing notifications to configured receivers via email, Slack, Telegram, and many others.
+
+To start creating Alertmanager alerts, you need to tell Uptrace how to reach AlertManager:
```
```suggestion
Uptrace does not manage notifications by itself and instead provides an [integration](https://uptrace.dev/get/alerting.html) with AlertManager.

AlertManager handles alerts sent by client applications such as Uptrace and takes care of deduplicating, grouping, and routing notifications to configured receivers via email, Slack, Telegram, and many others.

To connect Uptrace to AlertManager and enable alert notifications, add the AlertManager API endpoint to uptrace.yml:
```

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,189 @@
+
+Because installing OpenTelemetry Collector and Uptrace can take some time, you can use the [docker-compose](https://github.com/uptrace/uptrace/tree/master/example/redis-enterprise) example that also comes with Redis Enterprise cluster and Alertmanager.
+
+But you can also install OpenTelemetry and Uptrace from scratch using the following guides:
```
```suggestion
You can also install OpenTelemetry and Uptrace from scratch using the following guides:
```

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,189 @@
+
+## Monitor metrics
+
+To start monitoring metrics, you need to add some alerting rules, for example, the following rule creates an alert whenever an individual Redis shard is down which is ensured by the `group by node` expression:
```
```suggestion
To start monitoring metrics, you need to add some alerting rules.

For example, the following rule uses the `group by node` expression to create an alert whenever an individual Redis shard is down:
```

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
````diff
@@ -0,0 +1,189 @@
+      projects: [1]
+```
+
+You can also create queries with more complex expressions, for example, to alert when the keyspace hit rate is lower than 75% or memory fragmentation is too high:
````
```suggestion
You can also create queries with more complex expressions.

For example, the following rules create an alert when the keyspace hit rate is lower than 75% or memory fragmentation is too high:
```

rrelledge reviewed (changes requested):
Thanks for the updates, @vmihailenco. I left a few more small suggestions.

> The location is the same, but you can't edit files inside the docker container and that is why I have not specified the full path initially.

If the docker-compose installation method doesn't let you edit the config files, how do you change the alerts?
Do you just need to edit the config files in the cloned /uptrace/example/redis-enterprise/ directory before you run any docker-compose commands?

vmihailenco commented:
>Do you just need to edit the config files in the cloned /uptrace/example/redis-enterprise/ directory before you run any docker-compose commands?

Yes, exactly. I will add some explanation.

vmihailenco commented:
@rrelledge thanks for suggestions and please take another look.

rrelledge commented on content/rs/clusters/monitoring/uptrace-integration.md:
```diff
@@ -0,0 +1,198 @@
+
+Because installing OpenTelemetry Collector and Uptrace can take some time, you can use the [docker-compose](https://github.com/uptrace/uptrace/tree/master/example/redis-enterprise) example that also comes with Redis Enterprise cluster and AlertManager.
+
+The docker example contains the following files that can edit before starting Docker containers:
+
+- `otel-collector.yaml` - OpenTelemetry Collector config that you can find inside the running Collector container at `/etc/otelcol-contrib/config.yaml`.
+- `uptrace.yml` - Uptrace config that you can find inside the Uptrace container at `/etc/uptrace/uptrace.yml`.
```
```suggestion
After you download the Docker example, you can edit the following configuration files in the `uptrace/example/redis-enterprise` directory before you start the Docker containers:

- `otel-collector.yaml` - Configures `/etc/otelcol-contrib/config.yaml` in the OpenTelemetry Collector container.
- `uptrace.yml` - Configures`/etc/uptrace/uptrace.yml` in the Uptrace container.
```

rrelledge reviewed (changes requested):
Thank you for adding the Docker config details. This looks good to me.

vmihailenco commented:
@rrelledge just applied changes you suggested. Is anything else to do here?

rrelledge reviewed (approved):
Thanks, @vmihailenco. Looks good to me.

vmihailenco commented:
@rrelledge thanks :+1:

amiramm commented:
Hey @vmihailenco, thanks for this contribution. If folks want to get help with Uptrace, can we direct them to https://github.com/uptrace/uptrace#telegram ? I'd like to add a note about that into the article.

vmihailenco commented:
hey @amiramm , thanks and yes, but I've updated the link to https://github.com/uptrace/uptrace#help so it is more likely to survive. The contents is the same.
